docs: Add documentation for array type(s) (#110)
tableau/hyper-db: website/docs/sql/datatype/array.md

# Array Type

> An array is a sequential collection of elements packed into a single SQL value.

Reviewer: We should align our documentation with the one from PostgreSQL where feasiable and where we agree with the PostgreSQL documentation. It doesn't need to be exact, but I think the PostgreSQL array page has quite some good contents we should copy.

Author: I'm not sure I agree. I found the PostgreSQL documentation to be quite obtuse and difficult to navigate. Particularly, my gripes with it are:

 - The audience for this documentation are probably not software engineers, but business-analyst-type roles. The PostgreSQL docs never explain what an array is; they also refer to `C` without specifying it as a programming language. I don't think this is the level-of-detail we're aiming for.
 - In many places, the PostgreSQL docs introduce some concept or syntax (e.g., size specifiers, dimension specifiers) only to mention in the next sentence that the current implementation ignores it. Admittedly, this is probably more a testament to PostgreSQL's array design, but is also, again, a very developer-like view. In my experience, users do not have internalized the notion of interface vs. implementation; and they should not have to care. I do not think PostgreSQL's documentation shines in this regard.
 - Having explanations and examples interleaved (as PostgreSQL does) makes a good first-time reading experience, but makes it hard to locate information. When implementing arrays, I personally had a hard time looking up certain things, as speculative `CTRL`-`F `yields many false positives with this prose-heavy style. I would estimate that most readers are looking for a reference, rather than a tutorial.

Could you please be more specific in what you like about the PostgreSQL documentation, and why?

Author: See [below](https://github.com/tableau/hyper-db/pull/110#discussion_r1704373872): 
[bda7bea](https://github.com/tableau/hyper-db/pull/110/commits/bda7bea6e20cbb11681f1ca50140f8258c77a809)

Arrays are useful to model dependent information with an inherent ordering in a single row – for example, time-series data from longitudinal studies, or embedding vectors from machine learning models. 

> Generally, array values are denoted in curly braces with elements seperated by commas, like so:

Reviewer 2: ```suggestion
Generally, array values are denoted in curly braces with elements separated by commas, like so:
```

Author: [9fabdcd](https://github.com/tableau/hyper-db/pull/110/commits/9fabdcd8f563d54f052cc80085c08c87b54f630a)

> Generally, array values are denoted in curly braces with elements separated by commas, like so:

Reviewer: What means generally here? Where? See PostgreSQL documentation, they explain the syntax part better IMHO.

Author: (I agree with your remarks, about the leading "generally"; it's superfluous and confusing.)

See my comment above, could you perhaps point out what specifically you like about the PostgreSQL syntax documentation? 

The section on array input in the Postgres Docs refers to C struct initializers and has to introduce a generic format to accommodate `box[]` – I'm not sure that we should take it as a reference.

Reviewer: Also if you leave out "Generally", what means "are denoted in curly braces"? Denoted in this documentation? Denoted in Hyper SQL? If the latter, what part of it? E>g., I cannot just use curly braces in SQL, so this definitly doesn't seem to be SQL syntax. I can use them in strings, but that should be called out as such.

The [PostgreSQL documentation](https://www.postgresql.org/docs/current/arrays.html#ARRAYS-INPUT) for example mentions exactly how an array string literal has to look like, including how values need to be escaped.  This is missing here if I see it correctly.

Author: I reworked the section: [bda7bea](https://github.com/tableau/hyper-db/pull/110/commits/bda7bea6e20cbb11681f1ca50140f8258c77a809)

```

> {'Lorem', 'ipsum', 'dolor', 'sit', 'amed'}

Reviewer 2: Using double quotes for the string values inside the array allows users to copy paste this into a single quote escaped array literal.

```suggestion
{"Lorem", "ipsum", "dolor", "sit", "amed"}
```

Author: [9fabdcd](https://github.com/tableau/hyper-db/pull/110/commits/9fabdcd8f563d54f052cc80085c08c87b54f630a)

```

## Element Types and Nullability

Hyper's arrays are _strongly typed_:
All elements must be of the same type – the array's _element type_.

> The element type is defining part of the array's overall type, meaning that `array(integer)` is a different type than `array(boolean)` or `array(text)`.

Reviewer 2: ```suggestion
The element type defines part of the array's overall type, meaning that `array(integer)` is a different type than `array(boolean)` or `array(text)`.
```

Author: [9fabdcd](https://github.com/tableau/hyper-db/pull/110/commits/9fabdcd8f563d54f052cc80085c08c87b54f630a)

Arrays can be built from all [atomic types](./index.md) available in Hyper.

Part of an array's element type is its nullability. 

> For example, `array(smallint)` is a different array type than `array(smallint not null)`.

Reviewer: ```suggestion
For example, `array(smallint)` is different from `array(smallint not null)`.
```

Author: [52b62f2](https://github.com/tableau/hyper-db/pull/110/commits/52b62f29eae8f0581c35130b1d39ee3bb2dcf7d4)

> Note that this is independent of the array's overall nullability.

Reviewer: ```suggestion
Note that this is independent of the array type's overall nullability.
```

Author: [52b62f2](https://github.com/tableau/hyper-db/pull/110/commits/52b62f29eae8f0581c35130b1d39ee3bb2dcf7d4)

The following four options all represent different types in Hyper:

|Type|array nullable?|elements nullable?| possible values|
|---|---|---|---|
|`array(integer)`|✅|✅|`{}`,`{1,2,3}`,`{1,2,null}`, `null`|
|`array(integer not null)`|✅|❌|`{}`,`{1,2,3}`,`null`|
|`array(integer) not null`|❌|✅|`{}`,`{1,2,3}`,`{1,2,null}`|
|`array(integer not null) not null`|❌|❌|`{}`,`{1,2,3}`|

> Array types can be converted using the conventional [cast syntax](../scalar_func/conversion.md).

Author: [a5a69a0](https://github.com/tableau/hyper-db/pull/110/commits/a5a69a0e264b86285af85d5784cc4ef6e8783b10)

:::info

> Non-nullable element types use less memory and enable optimizations for certain array operatios. Users are therefore advised to use the most "restrictive" element type possible, if the use semantics of the case allows it. 

Reviewer 2: ```suggestion
Non-nullable element types use less memory and enable optimizations for certain array operations. Users are therefore advised to use the most "restrictive" element type possible, if the semantics of the use case allows it. 
```

Author: [9fabdcd](https://github.com/tableau/hyper-db/pull/110/commits/9fabdcd8f563d54f052cc80085c08c87b54f630a)

:::

> For nullable types, there exists an alternative shorthand bracket syntax of the form `type[]`. For example, `integer[]` and `array(integer)` reference the same type.

Reviewer: Should we even mention this? If so, should we mention it in a PostgreSQL compatibility section?

Author: > Should we even mention this?
I think we should definitely mention it, because this is the syntax returned by, e.g., `pg_typeof`.

> If so, should we mention it in a PostgreSQL compatibility section?
We could mention that the syntax exists for compatibility reasons, but not sure if it warrants its own section. Now that `array_lower`, `array_upper`, etc being undocumented, and Hyper's arrays deviating significantly from those in PostgreSQL, what else do you suggest would go in that section?

Reviewer: Do we document `pg_typeof`? If so, please then state this fact here. If not, that's yet another argument to either move this into a compatibility section or omit it add all.

I was suggesting a PostgreSQL compatibility section, because we do that in many other places of the documentation. Please skim the rest of the documentation and try to align to the general style here.

The reason why an own section is usually good is that it keeps all the "PostgreSQL compatibility cruft" out of the main reading path, so that readers who are not needing this compatibility (most of them don't) do not have to clutter their mind with this. People who just want to use arrays in Hyper do not need to be burdened with an extra syntax at all.

> Contrary to some common programming languages, the length (i.e., the number of elements) of an array in Hyper is not part of its type.

Reviewer: ```suggestion
The length (i.e., the number of elements) of an array in Hyper is not part of its type.
```

Author: [52b62f2](https://github.com/tableau/hyper-db/pull/110/commits/52b62f29eae8f0581c35130b1d39ee3bb2dcf7d4)

While arrays inside a column must have the same element type, they can be of different length.

## Working with Arrays

Arrays can be created in two ways:

 - Using the type constructor syntax:
   ```sql
   > select array[1,2,3];
   {1,2,3}
   ```
 - Using casts from string data:
   ```sql
   > select '{1,2,3}'::array(integer);
   {1,2,3}
   ```

Note that using the constructor syntax, the array type will be inferred automatically.
If the constructed array does not contain a `null` value, the element type will be inferred as non-nullable.

> Array elements can be retrieved using the conventional bracket-indexing notation. Indexes always start at one.

Reviewer: ```suggestion
Array elements can be retrieved using the conventional bracket-indexing notation. Indexes start at one.
```

Author: https://github.com/tableau/hyper-db/pull/110/commits/e5aaa9601d063f9281119ebe2cf36eeb8842eb9c

> Array elements can retrieved using the conventional bracket-indexing notation. Indexes always start at one.

Reviewer 2: ```suggestion
Array elements can be retrieved using the conventional bracket-indexing notation. Indexes always start at one.
```

Author: [9fabdcd](https://github.com/tableau/hyper-db/pull/110/commits/9fabdcd8f563d54f052cc80085c08c87b54f630a)

```sql
> select ('{1,1,2,3,5}'::integer[])[4]
3
```

:::info

> Arrays are always one-dimensional. Higher-dimensional objects (e.g., matrices) must be flattened explicitly.

Reviewer: ```suggestion
Arrays are one-dimensional. Higher-dimensional objects (e.g., matrices) must be flattened explicitly.
```

Author: [52b62f2](https://github.com/tableau/hyper-db/pull/110/commits/52b62f29eae8f0581c35130b1d39ee3bb2dcf7d4)

:::

For more operations on array, see the section on [Array Functions](../scalar_func/arrays.md).

## Limitations

> While arrays are first-class citizens in Hyper, they are subject to some moderate limitations:

Reviewer: ```suggestion
While arrays are first-class citizens in Hyper, they are subject to some limitations:
```

Author: [d47d337](https://github.com/tableau/hyper-db/pull/110/commits/d47d3379c6e66bf83bf4603475a8ede088e05a58)

> - The size of an array is limited to 4,294,967,296 elements. The actual limit may be lower, depending on the size(s) and nullability of its elements.

Reviewer: ```suggestion
- The size of an array is limited to 4,294,967,296 elements. The actual limit may be lower, depending on the size(s) and nullability of its elements.
```
I'm not sure we should mention this. Have we mentioned these limits for strings?

Reviewer: > I think it's important to convey that there is a limit. 
All databases have limits, everywhere. I agree that mentioning a limit is useful, but the usefulness of this limit is dubitable, as you will not be able to create such large arrays and even if you were, the only data type that can indeed take that many elements is `bool`, so the number isn't even correct for most data types.

Reviewer: I have suggested an edit. I think it is better to talk about byte size here, because that is the actual limit that applies. Yes, a user might not exactly know what how many array elements that is, but at least it is the actual enforced limit.

Author: [c3fc7d2](https://github.com/tableau/hyper-db/pull/110/commits/c3fc7d2d906330993048ab45fa132f00fce8197e)

- Arrays cannot be nested.

> - Arrays cannot be persisted.

Reviewer: ```suggestion
- Arrays cannot be persisted.
```
What does persisted mean here?
Be more explicit.

Author: The problem here is that there are some schema entries affected by this (e.g., views), which are themselves not properly documented.

I rewrote it to be more explicit without touching on these cases, wdyt?
[d47d337](https://github.com/tableau/hyper-db/pull/110/commits/d47d3379c6e66bf83bf4603475a8ede088e05a58)

> - Persisting arrays requires [file format version 4](../../hyper-api/hyper_process.md#version-4) or higher.

Author: [c5c2486](https://github.com/tableau/hyper-db/pull/110/commits/c5c24866c3c874065963c3910fb97bddebdaa276)

Generally, array values are denoted in curly braces with elements separated by commas, like so:

```
{"Lorem", "ipsum", "dolor", "sit", "amed"}
```

## Element Types and Nullability

Hyper's arrays are _strongly typed_:
All elements must be of the same type – the array's _element type_.
The element type is a defining part of the array's overall type, meaning that `array(integer)` is a different type than `array(boolean)` or `array(text)`.
Arrays can be built from all [atomic types](./index.md) available in Hyper.

Part of an array's element type is the element's nullability. 
For example, `array(smallint)` is different from `array(smallint not null)`.
Note that this is independent of the nullability of the array itself.
The following four options all represent different types in Hyper:

|Type|array nullable?|elements nullable?| possible values|
|---|---|---|---|
|`array(integer)`|✅|✅|`{}`,`{1,2,3}`,`{1,2,null}`, `null`|
|`array(integer not null)`|✅|❌|`{}`,`{1,2,3}`,`null`|
|`array(integer) not null`|❌|✅|`{}`,`{1,2,3}`,`{1,2,null}`|
|`array(integer not null) not null`|❌|❌|`{}`,`{1,2,3}`|

> The inner nullability of an array type can be changed by casting, using the conventional [cast syntax](../scalar_func/conversion.md):

Reviewer: Here we should mention which casts are allowed for arrays. I guess it's not only nullability changes, but also casts from and to text.

Author: [bda7bea](https://github.com/tableau/hyper-db/pull/110/commits/bda7bea6e20cbb11681f1ca50140f8258c77a809)

Casts across element types (e.g. from `array(integer not null)` to `array(bigint not null)`) are currently not supported.

:::info
Non-nullable element types use less memory and enable optimizations for certain array operations. Users are therefore advised to use the most "restrictive" element type possible, given the use case at hand. 
:::

The length (i.e., the number of elements) of an array in Hyper is not part of its type.
While arrays inside a column must have the same element type, they can be of different length.

## Working with Arrays

Arrays can be created in two ways:

 - Using the type constructor syntax:
   ```sql
   > select array[1,2,3];
   {1,2,3}
   ```
 - Using casts from string data:
   ```sql
   > select '{1,2,3}'::array(integer);
   {1,2,3}
   ```

Note that using the constructor syntax, the array type will be inferred automatically.
If the constructed array does not contain a `null` value, the element type will be inferred as non-nullable.

Array elements can be retrieved using the conventional bracket-indexing notation. Indexes start at one.
```sql

> > select ('{1,1,2,3,5}'::integer[])[4]

Reviewer: ```suggestion
> select ('{1,1,2,3,5}'::array(integer))[4]
```

Author: Removed this example and referred to array functions page instead [bda7bea](https://github.com/tableau/hyper-db/pull/110/commits/bda7bea6e20cbb11681f1ca50140f8258c77a809)

3
```

:::info
Arrays are one-dimensional. Higher-dimensional objects (e.g., matrices) must be flattened explicitly.

> :::

Reviewer: This info box doesn't seem to work. At least it doesn't work in[ the preview on git-hub](https://github.com/tableau/hyper-db/blob/lwolf-W-15542353-document-array-type/website/docs/sql/datatype/array.md#:~:text=Non%2Dnullable%20element%20types%20use%20less%20memory%20and%20enable).

Author: <img width="1031" alt="Screenshot 2024-08-07 at 09 53 38" src="https://github.com/user-attachments/assets/49607e99-e0cb-4fd3-8a6d-878205f50162">

> For more operations on array, see the section on [Array Functions](../scalar_func/arrays.md).

Reviewer: ```suggestion
For more operations on arrays, see the section on [Array Functions](../scalar_func/arrays.md).
```

Author: [6bfab43](https://github.com/tableau/hyper-db/pull/110/commits/6bfab4360d9e22870ebe3bfaf36be2f4dd40bde6)

# Array Type

An array is a sequential collection of elements packed into a single SQL value.
Arrays are useful to model dependent information with an inherent ordering in a single row – for example, time-series data from longitudinal studies, or embedding vectors from machine learning models. 

Generally, array values are denoted in curly braces with elements separated by commas, like so:

```
{"Lorem", "ipsum", "dolor", "sit", "amed"}
```

> 

Reviewer: Either use two or one space before captions. Align with rest of documentation.

Author: If this is important we should lint it. [2749843](https://github.com/tableau/hyper-db/pull/110/commits/27498430ec1738af4c4028b7ac560f1c70f06fb4)

## Element Types and Nullability

Hyper's arrays are _strongly typed_:
All elements must be of the same type – the array's _element type_.
The element type is a defining part of the array's overall type, meaning that `array(integer)` is a different type than `array(boolean)` or `array(text)`.
Arrays can be built from all [atomic types](./index.md) available in Hyper.

> Part of an array's element type is its nullability. 

Author: [2749843](https://github.com/tableau/hyper-db/pull/110/commits/27498430ec1738af4c4028b7ac560f1c70f06fb4)
